Specify project dependencies in PlatformIO config

gautamjain opened the pull request:
Hi Reed,

The project wasn't building after I cloned it.  I added two library dependencies to the PlatformIO config file to fix it:

https://platformio.org/lib/show/2032/ArduinoSimpleLogging
https://platformio.org/lib/show/5676/IotWebConf

gautamjain committed:
Specify project dependencies in PlatformIO config

reedtaylor commented:
Thrilled to hear that you've cloned it & thanks for catching the missing dependencies.  (I was shopping through a number of different libraries and adding them ad-hoc, basically never finalized.)

Speaking of things I never got around to ... I was in a huge rush when I originally wrote most of this code and as a consequence the documentation and readability are not what I'd usually strive for.  I'd be happy to do a sprint to improve things; let me know if it's annoying to work on what is currently present.

Also more that happy to take further PRs, talk about ideas, etc.!

gautamjain commented:
Thanks for all the work you put into this! Being able to warm up the DE1 while I'm half awake in bed is awesome.

The code looks great so far. Very straightforward. Although I haven't looked at the web server much. I want to spend more time understanding the serial protocol first.

I only made two other changes:

1. I wanted to see the commands the tablet was sending to the DE1.  So I added an option to forward serial data from the BT module to all controllers.

2. The web pages were taking very long to load from the ESP32 (it gets worse when there is a TCP/USB connection open). Almost 8-10 secs. This was because HTML/CSS/JPG resources were being served sequentially and the cache headers weren't set on them (so they were being fetched again and again for every page load). As a quick fix, I changed it to fetch bootstrap from a CDN and the background DE1 image from this github repo. Could probably do the same for `cover.css`, but I left it as is for now.

I can open PRs for these if you think they'd be helpful.

lddubeau commented:
No substantial comments on point 1. I think something like that was planned. (Actually, it sounds like the change that was put on the [`promiscuous-mode` branch](https://github.com/reedtaylor/plucky/tree/promiscuous-mode).)

The change suggested in point 2 has usability implications. There is a good discussion to have here, no matter the end result.

Usability: the change makes an internet connection necessary for proper operation because resources need to be downloaded from a CDN. In some use-case scenarios internet access is very difficult to obtain, if not impossible, or undesirable to have. For instance, situations where machines are brought in for a demo (trade show, espresso course, travelling coffee shop, etc). In theory, if the resources were loaded once, caching should take care of not hitting the internet, but that's too easy to mess up in practice in my opinion.

There are also security implications, but IIRC Plucky does not use HTTPS for serving files. So right now it is probably moot. If Plucky were to use HTTPS, then it would be better to serve everything from it.

Still, it seems to me some work could be done on improving the serving code. It seems to me that the proper caching headers (as you @gautamjain  point out) could be set, and there may be other optimizations that could be done to speed up the delivery of resources: CSS files could be combined into a single file (each request is relatively expensive), static files may be pre-compressed at build time in order to serve compressed answers while not paying the cost of having to recompress them all the time, etc.

I was also going to say that `bootstrap.css` can be trimmed down but that seems to be already done.

@reedtaylor I still have not reinstalled Plucky in my Daybreak. The constraints on my time went from very high to extreme, I'm afraid.

reedtaylor commented:
> I wanted to see the commands the tablet was sending to the DE1. So I added an option to forward serial data from the BT module to all controllers.

Exactly as LDD mentions I also have a branch that accomplishes this same thing, built for @lddubeau  (again -- no pressure :)  I am very much in the same situation regarding priorities!!!)... That branch is called [promiscuous mode](https://github.com/reedtaylor/plucky/tree/promiscuous-mode).  This branch was thrown together in a bit of haste, but it does work for me.  The change is not very complex :) though it can be a breaking change for downstream systems if they are not built to filter out the "informational" messages from the "broadcast messages".

If you want to create another PR with your change in it, I'd be more than happy to compare the two implementations, settle on an approach that works well for folks, and just merge that functionality into the master.

reedtaylor commented:
> The web pages were taking very long to load from the ESP32 (it gets worse when there is a TCP/USB connection open). Almost 8-10 secs. This was because HTML/CSS/JPG resources were being served sequentially and the cache headers weren't set on them (so they were being fetched again and again for every page load). As a quick fix, I changed it to fetch bootstrap from a CDN and the background DE1 image from this github repo. Could probably do the same for cover.css, but I left it as is for now.

Fixing the cache headers is a no-brainer.  I opened #32 to capture it.

Webserver performance is indeed garbage.  The solution I prefer there is #5 which migrates the entire network stack incl. webserver to asynctcp which uses the other core, so that network comms are not blocking any of the realtime implementation on the core currently running our main sketch.  I don;t think it'll be too hard to do, just a matter of when.

I made the conscious choice to not serve bootstrap via CDN for a variety of reasons, notably that I didn't want to require a working internet connection at all.   The additional points that @lddubeau mention are all very good as well.  But this could easily be made into a user-option.

That said, even bootstrap was a kind of arbitrary choice on my part; if somoene better versed in frameworks wants to suggest an alternative one or how to better accomplish the goals, I'l all ears!

gautamjain commented:
You guys are right. The UI should function properly when there is no Internet connectivity.  I didn't consider that when I linked to the CDN.

#5 + combined with a smaller CSS framework would definitely make a noticeable difference.  I've used [Skeleton](http://getskeleton.com/) in the past, but it hasn't been updated in a while and there's probably something better out. I don't have much experience with CSS frameworks either.

Thanks for pointing out the `promiscuous-mode` branch. I'll try that out this week.
